File: macOS/Renderer.m
@import MetalKit;

#import "spacetime.h"
#import "Renderer.h"
#import "ShaderTypes.h"

const int RENDERER_FRAMES_PER_SECOND = 60;

@implementation Renderer {
    View *_view;
    id<MTLDevice> _device;
    id<MTLRenderPipelineState> _pipelineState;
    id<MTLCommandQueue> _commandQueue;
    id<MTLBuffer> _vertexBuffer;

    uint2 _viewportSize;
    NSUInteger _numVertices;
}

- (instancetype)initWithView:(View *)view {
    self = [super init];
    if (self) {
        _view = view;
        _device = view.device;

        view.preferredFramesPerSecond = RENDERER_FRAMES_PER_SECOND;

        [self loadMetal:view];
    }

    return self;
}

+ (NSData *)generateVertexData
{
    const Vertex quadVertices[] = {
        // Pixel positions, RGBA colors
        {{-20, 20}, {1, 0, 0, 1}},
        {{20, 20}, {0, 0, 1, 1}},
        {{-20, -20}, {0, 1, 0, 1}},

        {{20, -20}, {1, 0, 0, 1}},
        {{-20, -20}, {0, 1, 0, 1}},
        {{20, 20}, {0, 0, 1, 1}},
    };
    const NSUInteger NUM_COLUMNS = 25;
    const NSUInteger NUM_ROWS = 15;
    const NSUInteger NUM_VERTICES_PER_QUAD = sizeof(quadVertices) / sizeof(Vertex);
    const float QUAD_SPACING = 50.0;

    NSUInteger dataSize = sizeof(quadVertices) * NUM_COLUMNS * NUM_ROWS;
    NSMutableData *vertexData = [[NSMutableData alloc] initWithLength:dataSize];

    Vertex* currentQuad = vertexData.mutableBytes;

    for (NSUInteger row = 0; row < NUM_ROWS; row++) {
        for (NSUInteger column = 0; column < NUM_COLUMNS; column++) {
            float2 upperLeftPosition;
            upperLeftPosition.x = ((-((float)NUM_COLUMNS) / 2.0) + column) * QUAD_SPACING + QUAD_SPACING/2.0;
            upperLeftPosition.y = ((-((float)NUM_ROWS) / 2.0) + row) * QUAD_SPACING + QUAD_SPACING/2.0;

            memcpy(currentQuad, &quadVertices, sizeof(quadVertices));

            for (NSUInteger vertexInQuad = 0; vertexInQuad < NUM_VERTICES_PER_QUAD; vertexInQuad++)
            {
                currentQuad[vertexInQuad].position += upperLeftPosition;
            }

            currentQuad += 6;
        }
    }
    return vertexData;
}

- (void)loadMetal:(MTKView *)mtkView {
    mtkView.colorPixelFormat = MTLPixelFormatBGRA8Unorm_sRGB;

    id<MTLLibrary> defaultLibrary = [_device newDefaultLibrary];

    id<MTLFunction> vertexFunction = [defaultLibrary newFunctionWithName:@"vertexShader"];
    id<MTLFunction> fragmentFunction = [defaultLibrary newFunctionWithName:@"fragmentShader"];

    MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
    pipelineStateDescriptor.label = @"Simple Pipeline";
    pipelineStateDescriptor.vertexFunction = vertexFunction;
    pipelineStateDescriptor.fragmentFunction = fragmentFunction;
    pipelineStateDescriptor.colorAttachments[0].pixelFormat = mtkView.colorPixelFormat;

    NSError *error = NULL;
    _pipelineState = [_device newRenderPipelineStateWithDescriptor:pipelineStateDescriptor
                                                             error:&error];
    if (!_pipelineState) {
        // Pipeline State creation could fail if we haven't properly set up our pipeline descriptor.
        //  If the Metal API validation is enabled, we can find out more information about what
        //  went wrong.  (Metal API validation is enabled by default when a debug build is run
        //  from Xcode)
        NSLog(@"Failed to created pipeline state, error %@", error);
    }

    NSData *vertexData = [Renderer generateVertexData];

    // Create a vertex buffer by allocating storage that can be read by the GPU
    _vertexBuffer = [_device newBufferWithLength:vertexData.length
                                         options:MTLResourceStorageModeShared];

    // Copy the vertex data into the vertex buffer by accessing a pointer via
    // the buffer's `contents` property
    memcpy(_vertexBuffer.contents, vertexData.bytes, vertexData.length);

    // Calculate the number of vertices by dividing the byte length by the size of each vertex
    _numVertices = vertexData.length / sizeof(Vertex);

    _commandQueue = [_device newCommandQueue];
}

- (void)mtkView:(MTKView *)view drawableSizeWillChange:(CGSize)size {
    (void)view;

    // Save the size of the drawable as we'll pass these
    //   values to our vertex shader when we draw
    _viewportSize.x = size.width;
    _viewportSize.y = size.height;
}

- (void)render: {
    MTKView *view = _view;
    id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];
    commandBuffer.label = @"MyCommand";

    MTLRenderPassDescriptor *renderPassDescriptor = view.currentRenderPassDescriptor;

    if (renderPassDescriptor != nil) {
        id<MTLRenderCommandEncoder> renderEncoder =
        [commandBuffer renderCommandEncoderWithDescriptor:renderPassDescriptor];
        renderEncoder.label = @"MyRenderEncoder";

        // Set the region of the drawable to which we'll draw.
        [renderEncoder setViewport:(MTLViewport){0.0, 0.0, _viewportSize.x, _viewportSize.y, -1.0, 1.0 }];

        [renderEncoder setRenderPipelineState:_pipelineState];

        // We call -[MTLRenderCommandEncoder setVertexBuffer:offset:atIndex:] to send data in our
        //   preloaded MTLBuffer from our ObjC code here to our Metal 'vertexShader' function
        // This call has 3 arguments
        //   1) buffer - The buffer object containing the data we want passed down
        //   2) offset - They byte offset from the beginning of the buffer which indicates what
        //      'vertexPointer' point to.  In this case we pass 0 so data at the very beginning is
        //      passed down.
        //      We'll learn about potential uses of the offset in future samples
        //   3) index - An integer index which corresponds to the index of the buffer attribute
        //      qualifier of the argument in our 'vertexShader' function.  Note, this parameter is
        //      the same as the 'index' parameter in
        //              -[MTLRenderCommandEncoder setVertexBytes:length:atIndex:]
        //
        [renderEncoder setVertexBuffer:_vertexBuffer
                                offset:0
                               atIndex:VertexInputIndexVertices];

        [renderEncoder setVertexBytes:&_viewportSize
                               length:sizeof(_viewportSize)
                              atIndex:VertexInputIndexViewportSize];

        // Draw the vertices of the quads
        [renderEncoder drawPrimitives:MTLPrimitiveTypeTriangle
                          vertexStart:0
                          vertexCount:_numVertices];

        [renderEncoder endEncoding];

        [commandBuffer presentDrawable:view.currentDrawable];
    }

    [commandBuffer commit];
}

- (void)drawInMTKView:(MTKView *)view {
    @autoreleasepool {
        world_loop_tick(self);
    }
}

@end

File: src/Render/Renderer.c
#include "Render/Renderer.h"

#include "Gpu/Cmd.h"
#include "Gpu/Primitive.h"
#include "Gpu/ShaderTypes.h"

#include <stdlib.h>
#include <string.h>

void RendererInit(Renderer *renderer, GpuDevice *device, RendererConfig *config)
{
    renderer->vertex_function =
            GpuFunctionCreateWithName(device, "vertex_shader");
    renderer->fragment_function =
            GpuFunctionCreateWithName(device, "fragment_shader");

    GpuPipelineStateConfig pipeline_state_config = {
        .label = "Simple Pipeline",
        .vertex_function = renderer->vertex_function,
        .fragment_function = renderer->fragment_function,
        .pixel_format = config->view.color_pixel_format,
    };

    renderer->pipeline_state =
            GpuPipelineStateCreate(device, &pipeline_state_config);

    // Generate vertex buffer
    {
        const GpuVertex QuadVertices[] = {
            // Pixel positions, RGBA colors
            { { -20, 20 }, { 1, 0, 0, 1 } },  { { 20, 20 }, { 0, 0, 1, 1 } },
            { { -20, -20 }, { 0, 1, 0, 1 } },

            { { 20, -20 }, { 1, 0, 0, 1 } },  { { -20, -20 }, { 0, 1, 0, 1 } },
            { { 20, 20 }, { 0, 0, 1, 1 } },
        };
        const i64 NumRows = 15;
        const i64 NumColumns = 25;
        const i64 NumVerticesPerQuad = sizeof(QuadVertices) / sizeof(GpuVertex);
        const float QuadSpacing = 50.0;
        const i64 DataLength = sizeof(QuadVertices) * NumRows * NumColumns;

        renderer->vertex_buffer = GpuBufferCreateWithLength(
                device, DataLength, GpuBufferStorageModeShared);
        renderer->num_vertices = DataLength / sizeof(GpuVertex);

        GpuVertex *current_quad = GpuBufferContents(renderer->vertex_buffer);

        for (i32f row = 0; row < NumRows; row++) {
            for (i32f column = 0; column < NumColumns; column++) {
                float2 upper_left_position;
                upper_left_position.x =
                        ((-((float)NumColumns) / 2.0) + column) * QuadSpacing +
                        QuadSpacing / 2.0;
                upper_left_position.y =
                        ((-((float)NumRows) / 2.0) + row) * QuadSpacing +
                        QuadSpacing / 2.0;

                memcpy(current_quad, &QuadVertices, sizeof(QuadVertices));

                for (i32f vertex_in_quad = 0;
                     vertex_in_quad < NumVerticesPerQuad;
                     vertex_in_quad++) {
                    GpuVertex *current_vertex = &current_quad[vertex_in_quad];
                    current_vertex->position.x += upper_left_position.x;
                    current_vertex->position.y += upper_left_position.y;
                }

                current_quad += NumVerticesPerQuad;
            }
        }
    }
}

void RendererDestroy(Renderer *renderer)
{
    GpuBufferDestroy(renderer->vertex_buffer);
    GpuPipelineStateDestroy(renderer->pipeline_state);
    GpuFunctionDestroy(renderer->fragment_function);
    GpuFunctionDestroy(renderer->vertex_function);
}

void RendererDrawInView(Renderer *renderer,
                        GpuView *view,
                        RenderViewport *viewport,
                        GpuCommandQueue *queue)
{
    GpuCommandBuffer *command_buffer = GpuCommandBufferCreate(queue);
    GpuRenderPassConfig *render_pass_config =
            GpuViewCurrentRenderPassConfig(view);

    if (render_pass_config) {
        GpuCommandEncoder *encoder;
        GpuCommandEncoderBeginRenderEncoding(
                &encoder, command_buffer, render_pass_config);
        GpuCommandEncoderLabel(encoder, "MyRenderEncoder");

        GpuCmdSetViewport(encoder, viewport->size);

        GpuCmdSetPipelineState(encoder, renderer->pipeline_state);

        GpuCmdBindVertexBuffer(
                encoder, renderer->vertex_buffer, 0, GpuVertexIndexVertices);

        GpuCmdBindVertexBytes(encoder,
                              &viewport->size,
                              sizeof(viewport->size),
                              GpuVertexIndexViewportSize);

        GpuCmdDrawPrimitives(
                encoder, GpuPrimitiveTypeTriangle, 0, renderer->num_vertices);

        GpuCommandEncoderEndEncoding(encoder);
        GpuCommandBufferPresentDrawable(command_buffer, view);
        GpuRenderPassConfigDestroy(render_pass_config);
    }

    GpuCommandBufferCommit(command_buffer);
    GpuCommandBufferDestroy(command_buffer);
}
